INT-B-21505 prime port update

brianmanley-caci opened the pull request:
## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A1018852&RoomContext=TeamRoom%3A1007470&concept=TeamRoom)

## Summary
This story is to allow the prime user to update the port location for a move with either a POEFSC or PODFSC service item.

Is there anything you would like reviewers to give additional scrutiny?

[this article](tbd) explains more about the approach used.

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test
Note: the functionality for ordering POEFSC and PODFSC service items does not yet exist. Therefore in order to test you will need to manually create either a POEFSC or PODFSC service item for a shipment.

1. Create an INTL UB shipment as a customer (CONUS -> AK or AK -> CONUS)
2. Manually create the mto_service_items table for the shipment: `
INSERT INTO mto_service_items (
					id,
                    mto_shipment_id,
                    move_id,
                    re_service_id,
                    status,
                    created_at,
                    updated_at,
                    approved_at
                )
                VALUES (
                uuid_generate_v4(),
                    '8eeba537-bb40-4f58-a534-74ea9ff64590',
                    '813b7a43-2e3f-42f3-8c4e-eca95390d5ca',
                    '388115e8-abe9-441d-96cf-a39f24baa0a3',
                    'APPROVED'::service_item_status,
                    NOW(),
                    NOW(),
                    NOW()
                );
`
3. In the prime v3 api, call getMoveTaskOrder using the move_id for the service item you created in step 2. You will need the etag for the newly created service item for the next step.
4. In the prime v1 api, call updateMTOServiceItem using a payload like this (adjust for you specific service item ID): `
{
  "id": "40b7e71a-6b03-4413-8601-dea694253ee3",
  "modelType": "UpdateMTOServiceItemInternationalPortFSC",
  "portCode": "PDX",
  "reServiceCode": "PODFSC"
}
`
5. Verify that you receive a 200 response like this: `{
  "reServiceCode": "PODFSC",
  "eTag": "MjAyNC0xMi0wMlQxNjo0MjoxNi42NjUwNDla",
  "id": "40b7e71a-6b03-4413-8601-dea694253ee3",
  "modelType": "MTOServiceItemBasic",
  "moveTaskOrderID": "813b7a43-2e3f-42f3-8c4e-eca95390d5ca",
  "mtoShipmentID": "8eeba537-bb40-4f58-a534-74ea9ff64590",
  "reServiceName": "International POD Fuel Surcharge",
  "status": "APPROVED"
}`
6. In the prime v3 api, call getMoveTaskOrder using the same move_id as before.
7. Verify that the updated shipment now has the portOfDebarkation (or portOfEmbarkation) object as part of the response like this: 
`"portOfDebarkation": {
        "city": "PORTLAND",
        "country": "UNITED STATES",
        "county": "MULTNOMAH",
        "id": "b6e94f5b-33c0-43f3-b960-7c7b2a4ee5fc",
        "portCode": "PDX",
        "portName": "PORTLAND INTL",
        "portType": "A",
        "state": "OREGON",
        "zip": "97220"
      },`


### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

### Database

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

## Screenshots

danieljordan-caci commented on swagger-def/definitions/Port.yaml:
```diff
@@ -0,0 +1,95 @@
+description: A port
```
Can you add just a liiiiiittle bit more to this description?

brianmanley-caci replied:
Reasonable ask.

danieljordan-caci commented on swagger-def/definitions/Port.yaml:
```diff
@@ -0,0 +1,95 @@
+    example: 'Anchorage'
+  county:
+    type: string
+    example: 'Anchorage'
```
Please use another example to avoid having the same `Anchorage` example for different fields

brianmanley-caci replied:
Updated to example with greater variance.

danieljordan-caci commented on swagger-def/definitions/Port.yaml:
```diff
@@ -0,0 +1,95 @@
+  portName:
+    type: string
+    description: 'Name of the port'
+    example: 'Anchorage'
```
Please use another example to avoid having the same `Anchorage` example for different fields

brianmanley-caci replied:
Updated to example with greater variance.

danieljordan-caci reviewed (changes requested):
Static review for now

brianmanley-caci commented on pkg/handlers/primeapiv3/payloads/model_to_payload.go:
```diff
@@ -1073,3 +1075,46 @@ func PostalCodeToRateArea(postalCode *string, shipmentPostalCodeRateAreaMap map[
+	shipmentPodMap := make(map[string]*models.PortLocation)
+	shipmentPoeMap := make(map[string]*models.PortLocation)
+	for _, mtoServiceItem := range *mtoServiceItems {
+		if mtoServiceItem.PODLocation != nil {
+			shipmentPodMap[mtoServiceItem.MTOShipmentID.String()] = mtoServiceItem.PODLocation
+		} else if mtoServiceItem.POELocation != nil {
+			shipmentPoeMap[mtoServiceItem.MTOShipmentID.String()] = mtoServiceItem.POELocation
+		}
+	}
+	var podMapEmpty = len(shipmentPodMap) == 0
+	var poeMapEmpty = len(shipmentPoeMap) == 0
+	if !podMapEmpty || !poeMapEmpty {
+		for _, mtoShipment := range *mtoShipments {
+			if !podMapEmpty && shipmentPodMap[string(mtoShipment.ID)] != nil {
+				podLocation := shipmentPodMap[string(mtoShipment.ID)]
+				pod := getPortMessageFromPortLocation(podLocation)
+				mtoShipment.PortOfDebarkation = pod
+			} else if !poeMapEmpty && shipmentPoeMap[string(mtoShipment.ID)] != nil {
+				poeLocation := shipmentPoeMap[string(mtoShipment.ID)]
+				poe := getPortMessageFromPortLocation(poeLocation)
+				mtoShipment.PortOfEmbarkation = poe
+			}
+		}
+	}
+}
+
+// Convert a PortLocation model to Port message
+func getPortMessageFromPortLocation(portLocation *models.PortLocation) *primev3messages.Port {
+	return &primev3messages.Port{
+		ID:       strfmt.UUID(portLocation.ID.String()),
+		PortType: portLocation.Port.PortType.String(),
+		PortCode: portLocation.Port.PortCode,
+		PortName: portLocation.Port.PortName,
+		City:     portLocation.City.CityName,
+		County:   portLocation.UsPostRegionCity.UsprcCountyNm,
+		State:    portLocation.UsPostRegionCity.UsPostRegion.State.StateName,
+		Zip:      portLocation.UsPostRegionCity.UsprZipID,
+		Country:  portLocation.Country.CountryName,
+	}
+}
```
I can rename the function. I separated it out to try to avoid duplicating the logic if that makes sense.

brianmanley-caci replied:
The function is returning a port based off the port location, so would it make more sense to name the function Port?

danieljordan-caci replied:
Oh gotcha! Missed. Then yes! I think that'd be good!

brianmanley-caci commented on pkg/services/mto_service_item/mto_service_item_updater_test.go:
```diff
@@ -1429,6 +1431,206 @@ func (suite *MTOServiceItemServiceSuite) TestValidateUpdateMTOServiceItem() {
+		suite.Equal(updatedServiceItem.POELocationID, newServiceItemPrime.POELocationID)
+	})
+
+	// Test failure Prime validation for Port of Embarkation
```
Good catch. Copy paste error.

brianmanley-caci replied:
Fixed.

brianmanley-caci commented on pkg/services/mto_service_item/mto_service_item_updater_test.go:
```diff
@@ -1429,6 +1431,206 @@ func (suite *MTOServiceItemServiceSuite) TestValidateUpdateMTOServiceItem() {
+		suite.Equal(updatedServiceItem.PODLocationID, newServiceItemPrime.PODLocationID)
+	})
+
+	// Test failure Prime validation for Port of Debarkation
```
Good catch. Copy paste error.

brianmanley-caci replied:
Fixed.

brianmanley-caci commented on pkg/handlers/primeapiv3/payloads/model_to_payload_test.go:
```diff
@@ -1284,3 +1284,145 @@ func (suite *PayloadsSuite) TestMarketCode() {
+		Country: models.Country{
+			CountryName: country,
+		},
+	}
```
I created a port and port location factory. Let me know if it is any better?

brianmanley-caci commented on pkg/handlers/primeapiv3/payloads/model_to_payload_test.go:
```diff
@@ -1284,3 +1284,145 @@ func (suite *PayloadsSuite) TestMarketCode() {
+	zip := "97220"
+	country := "US"
+	reServiceCode := models.ReServiceCodePOEFSC
+	mtoShipmentId := uuid.Must(uuid.NewV4())
```
Removed uuid generator and used factories.

brianmanley-caci commented on pkg/handlers/primeapiv3/payloads/model_to_payload_test.go:
```diff
@@ -1284,3 +1284,145 @@ func (suite *PayloadsSuite) TestMarketCode() {
+		Country: models.Country{
+			CountryName: country,
+		},
+	}
```
refactored to use factories.

brianmanley-caci commented:
@danieljordan-caci ready for further scrutiny.

danieljordan-caci commented on pkg/factory/port_location_factory.go:
```diff
@@ -0,0 +1,62 @@
+		}
+	}
+
+	defaultPortLocation := models.PortLocation{
+		Port: models.Port{
+			PortType: models.PortTypeAir,
+			PortCode: "PDX",
+			PortName: "PORTLAND INTL",
+		},
+		City: models.City{
+			CityName: "PORTLAND",
+		},
+		UsPostRegionCity: models.UsPostRegionCity{
+			UsprcCountyNm: "MULTNOMAH",
+			UsprZipID:     "97220",
+			UsPostRegion: models.UsPostRegion{
+				State: models.State{
+					StateName: "OREGON",
+				},
+			},
+		},
+		Country: models.Country{
+			CountryName: "UNITED STATES",
+		},
+	}
```
So for this part, since we want to return an ACTUAL row from `port_locations`, instead of returning default data you should instead fetch the default one since the user isn't specifying which one. Looking better! 

```
    var portLocation models.PortLocation
	if db != nil {
		err := db.EagerPreload("Port", "City", "Country", "UsPostRegionCity.UsPostRegion.State").Where("is_active = TRUE").InnerJoin("ports p", "port_id = p.id").Where("p.port_code = $1", port.PortCode).First(&portLocation)
		if err == nil {
			return portLocation
		}
	}

	// search for the default port if one is not provided
	defaultPortCode := "PDX"
	err := db.EagerPreload("Port", "City", "Country", "UsPostRegionCity.UsPostRegion.State").Where("is_active = TRUE").InnerJoin("ports p", "port_id = p.id").Where("p.port_code = $1", port.PortCode).First(&defaultPortCode)
	if err != nil && err != sql.ErrNoRows {
		log.Panic(err)
	} else {
		return portLocation
	}
```

brianmanley-caci replied:
Updated to pull default record from the DB.

danieljordan-caci replied:
Same here - not seeing these changes. You push your changes?

brianmanley-caci replied:
Last week was a blur. I can see the changes in there now. Let me know if you are still not seeing them.

danieljordan-caci commented on pkg/factory/mto_service_item_factory.go:
```diff
@@ -55,6 +55,24 @@ func buildMTOServiceItemWithBuildType(db *pop.Connection, customs []Customizatio
 		reService = FetchReServiceByCode(db, models.ReServiceCode("DLH"))
 	}
 
+	var ptrPodLocation *models.PortLocation
+	if result := findValidCustomization(customs, PortLocations.PortOfDebarkation); result != nil {
+		podLocation := result.Model.(models.PortLocation)
+		if !result.LinkOnly {
+			podLocation = FetchPortLocation(db, customs, nil)
+		}
+		ptrPodLocation = &podLocation
+	}
+
+	var ptrPoeLocation *models.PortLocation
+	if result := findValidCustomization(customs, PortLocations.PortOfEmbarkation); result != nil {
+		poeLocation := result.Model.(models.PortLocation)
+		if !result.LinkOnly {
+			poeLocation = FetchPortLocation(db, customs, nil)
+		}
+		ptrPoeLocation = &poeLocation
+	}
+
```
Can you add/update a test in `mto_service_item_factory_test.go` to reflect these changes?

brianmanley-caci replied:
Added a test to validate the ports in this function.

danieljordan-caci replied:
Not seeing a change in `mto_service_item_factory_test.go`?

brianmanley-caci replied:
just merged and pushed back up. It should be up to date now.

danieljordan-caci replied:
Thanks for the test!

danieljordan-caci commented on pkg/factory/port_factory.go:
```diff
@@ -0,0 +1,38 @@
+		}
+	}
+
+	port := models.Port{
+		PortCode: "PDX",
+		PortName: "PORTLAND INTL",
+		PortType: "A",
+	}
```
So same thing here as the port locations factory - fetch a default port instead of returning dummy data.

brianmanley-caci replied:
Grabbing default from the db.

danieljordan-caci replied:
Not seeing this change?

brianmanley-caci replied:
just merged and pushed back up. It should be up to date now.

danieljordan-caci commented on pkg/handlers/primeapi/payloads/payload_to_model_test.go:
```diff
@@ -566,6 +566,36 @@ func (suite *PayloadsSuite) TestMTOServiceItemModelFromUpdate() {
+			PortCode:      &portCode,
+		}
+
+		model, _ := MTOServiceItemModelFromUpdate(mtoServiceItemID, &updateMTOServiceInternationalPortFsc)
```
Check for no error:
```
suite.NoError(err)
```

brianmanley-caci replied:
validated errors is empty.

danieljordan-caci commented on pkg/handlers/primeapi/payloads/payload_to_model_test.go:
```diff
@@ -566,6 +566,36 @@ func (suite *PayloadsSuite) TestMTOServiceItemModelFromUpdate() {
+			PortCode:      &portCode,
+		}
+
+		model, _ := MTOServiceItemModelFromUpdate(mtoServiceItemID, &updateMTOServiceInternationalPortFsc)
```
Check for no error:
```
suite.NoError(err)
```

brianmanley-caci replied:
validated errors is empty

danieljordan-caci commented on swagger-def/definitions/Port.yaml:
```diff
@@ -0,0 +1,95 @@
+    example: c56a4180-65aa-42ec-a945-5fd21dec0538
+  portType:
+    type: string
+    description: 'Port type A (Air), B(Border Crossing), S (Sea)'
```
Add a space after `B` here to keep up with what you're doing for `A` and `S`
```
'Port type A (Air), B (Border Crossing), S (Sea)'
```

brianmanley-caci replied:
added space for consistency.

danieljordan-caci reviewed (changes requested):
Closer! Just a few more changes pleaseeeee

robot-mymove commented:
# Test coverage on server failed.
  **Last run: Wed 04 Dec 2024 05:28:28 PM UTC**
  Refer to [this report](https://output.circle-artifacts.com/output/job/402132b2-f4cd-433e-9bab-4f3fcbcce57b/artifacts/0/test-results/gotest/app/go-coverage.html) to see coverage details for the server app.
  Documentation for test coverage can be found [at this link](https://transcom.github.io/mymove-docs/docs/about/development/test-coverage-reports).

WeatherfordAaron commented on pkg/services/move_task_order/move_task_order_fetcher.go:
```diff
@@ -210,6 +214,21 @@ func (f moveTaskOrderFetcher) FetchMoveTaskOrder(appCtx appcontext.AppContext, s
+				return &models.Move{}, apperror.NewQueryError("PODLocation", loadErr, "")
+			}
+		}
+		if serviceItem.POELocation != nil {
```
I'd consider combining these 2 if blocks since they both look for the same condition.

danieljordan-caci replied:
Doesn't look like the logic after the `if` is the same, so I would say this approach is fine. If `PODLocation` isn't `nil`, then he's querying for the `PODLocation` on the service item. Same for `POELocation`.

WeatherfordAaron replied:
You are correct. I missed the last letters of those.

danieljordan-caci commented on pkg/factory/port_location_factory.go:
```diff
@@ -0,0 +1,47 @@
+		}
+
+		// Find the port location based on the port code
+		err := db.EagerPreload("Port", "City", "Country", "UsPostRegionCity.UsPostRegion.State").Where("is_active = TRUE").InnerJoin("ports p", "port_id = p.id").Where("p.port_code = $1", port.PortCode).First(&portLocation)
+		if err == nil {
+			return portLocation
+		}
+
+		// Didn't find a port location based on the custom port code, so grab the default port location
+		err = db.EagerPreload("Port", "City", "Country", "UsPostRegionCity.UsPostRegion.State").Where("is_active = TRUE").InnerJoin("ports p", "port_id = p.id").Where("p.port_code = 'PDX'").First(&portLocation)
+		if err == nil {
+			return portLocation
+		}
```
You want to do an `if` check here. Could even do an `if/else` check.

```
if port.PortCode != nil {
...fetch the requested port location
}
else {
...fetch default port location
}

brianmanley-caci replied:
Updated to check for empty since nil is not valid.

danieljordan-caci commented on pkg/factory/port_factory.go:
```diff
@@ -0,0 +1,38 @@
+	}
+
+	var port models.Port
+	if db != nil {
+		err := db.Where("port_code = ?", cPort.PortCode).First(&port)
+		if err == nil {
+			return port
+		}
+
+		// Didn't find a port based on the custom port code, so grab the default port
+		err = db.Where("port_code = 'PDX'").First(&port)
+		if err == nil {
+			return port
+		}
+	}
```
Same thing here - do a `if cPort.PortCode != nil` check, if not, fetch the default.

brianmanley-caci replied:
Updated to check for empty since nil is not valid here too.

danieljordan-caci reviewed (changes requested):
Just one more little change with your factories. Rest look good.

msaki-caci commented:
> Anybody else wanna review this lol

I have been trying to today but I can't UB to work haha

msaki-caci commented on pkg/factory/shared.go:
```diff
@@ -275,6 +279,18 @@ var Uploads = uploadGroup{
 	UploadTypeUser:  "UploadTypeUser",
 }
 
+// portGroup is a grouping of all the port related fields
```
Should this be "portLocationGroup is a grouping...."?

brianmanley-caci replied:
Yep. Fixed.

msaki-caci commented on pkg/services/move_task_order/move_task_order_fetcher_test.go:
```diff
@@ -391,6 +391,86 @@ func (suite *MoveTaskOrderServiceSuite) TestMoveTaskOrderFetcher() {
+			}
+		}
+		// Verify that the expected service item was found
+		suite.True(found, "Expected service item ReServiceCodePOEFSC")
```
Should this be "Expected service item **ReServiceCodePODFSC**"

brianmanley-caci replied:
Updated to correct code.

msaki-caci commented on pkg/services/mto_service_item/mto_service_item_updater_test.go:
```diff
@@ -1429,6 +1431,206 @@ func (suite *MTOServiceItemServiceSuite) TestValidateUpdateMTOServiceItem() {
+		suite.Equal(oldServiceItemPrime.PODLocationID, updatedServiceItem.PODLocationID)
+	})
+
+	// Test failure Prime validation for Port of Embarkation
```
Should this be "Port of Debarkation"?

brianmanley-caci replied:
Updated to correct wording.

robot-mymove commented:
# Test coverage on server failed.
  **Last run: Wed 11 Dec 2024 08:54:29 PM UTC**
  Refer to [this report](https://output.circle-artifacts.com/output/job/ed77e7ba-a27e-417e-a923-caa76d521bba/artifacts/0/test-results/gotest/app/go-coverage.html) to see coverage details for the server app.
  Documentation for test coverage can be found [at this link](https://transcom.github.io/mymove-docs/docs/about/development/test-coverage-reports).

msaki-caci commented on yarn.lock:
On your latest commit this file was modified. Was this on purpose?

brianmanley-caci replied:
That change came from main. @cameroncaci can you confirm that this is correct best off your recent main PR for B-21230?

msaki-caci replied:
I will go ahead and resolve this conversation

robot-mymove commented:
# Test coverage on server failed.
  **Last run: Wed 11 Dec 2024 11:51:13 PM UTC**
  Refer to [this report](https://output.circle-artifacts.com/output/job/0d71daa8-e454-46ea-9e17-57f3aa477ac5/artifacts/0/test-results/gotest/app/go-coverage.html) to see coverage details for the server app.
  Documentation for test coverage can be found [at this link](https://transcom.github.io/mymove-docs/docs/about/development/test-coverage-reports).

danieljordan-caci reviewed (commented):
Question - was able to update the service item, but here's my payload:
```
{   
   "id": "ad1dc99c-3c51-4c0b-94eb-ace44d6be2f1",   
   "modelType": "UpdateMTOServiceItemInternationalPortFSC",   
   "portCode": "PDX",   
   "reServiceCode": "POEFSC" 
}
```

And here's my response:
```
{
  "reServiceCode": "POEFSC",
  "eTag": "MjAyNC0xMi0xMlQxNDo1NDo0OS40MzQ4MDJa",
  "id": "ad1dc99c-3c51-4c0b-94eb-ace44d6be2f1",
  "modelType": "MTOServiceItemBasic",
  "moveTaskOrderID": "5a99af3f-e9ce-4973-a064-b1606e3d95e8",
  "mtoShipmentID": "113a4bc3-ad46-464c-a98f-396b65ea2872",
  "reServiceName": "International POE Fuel Surcharge",
  "status": "APPROVED"
}
```

I just am not sure how I feel about us not returning the values that we have updated to confirm that they have been updated correctly. Here I updated the port, but I have no way of knowing it was successful (other than the `200`) in the response. Any thoughts on that? We could entertain having a `MTOServiceItemFuelSurcharge` response that contains that data, or add it to `MTOServiceItemBasic`. 

But it's also not a hill I want to die on and if the gov is cool with the response lacking all the port info, then I don't mind approving.

danieljordan-caci reviewed (changes requested):
Ah, but while I'm thinking of it - please update the docs for the update service item endpoint on what is required when updating `PODFSC` & `POEFSC`. Just a brief run through of our process, what we require, etc, so we can inform HomeSafe when they start implementing this.

brianmanley-caci commented:
> Question - was able to update the service item, but here's my payload:
> 
> ```
> {   
>    "id": "ad1dc99c-3c51-4c0b-94eb-ace44d6be2f1",   
>    "modelType": "UpdateMTOServiceItemInternationalPortFSC",   
>    "portCode": "PDX",   
>    "reServiceCode": "POEFSC" 
> }
> ```
> 
> And here's my response:
> 
> ```
> {
>   "reServiceCode": "POEFSC",
>   "eTag": "MjAyNC0xMi0xMlQxNDo1NDo0OS40MzQ4MDJa",
>   "id": "ad1dc99c-3c51-4c0b-94eb-ace44d6be2f1",
>   "modelType": "MTOServiceItemBasic",
>   "moveTaskOrderID": "5a99af3f-e9ce-4973-a064-b1606e3d95e8",
>   "mtoShipmentID": "113a4bc3-ad46-464c-a98f-396b65ea2872",
>   "reServiceName": "International POE Fuel Surcharge",
>   "status": "APPROVED"
> }
> ```
> 
> I just am not sure how I feel about us not returning the values that we have updated to confirm that they have been updated correctly. Here I updated the port, but I have no way of knowing it was successful (other than the `200`) in the response. Any thoughts on that? We could entertain having a `MTOServiceItemFuelSurcharge` response that contains that data, or add it to `MTOServiceItemBasic`.
> 
> But it's also not a hill I want to die on and if the gov is cool with the response lacking all the port info, then I don't mind approving.

Yeah I do see what you're saying. Do you know other service items return any kind of info in the response like when they do an update? The way this feature spelled it out, after sending this update they would subsequently query the MTO and see the port information in that payload.

danieljordan-caci commented:
> > Question - was able to update the service item, but here's my payload:
> > ```
> > {   
> >    "id": "ad1dc99c-3c51-4c0b-94eb-ace44d6be2f1",   
> >    "modelType": "UpdateMTOServiceItemInternationalPortFSC",   
> >    "portCode": "PDX",   
> >    "reServiceCode": "POEFSC" 
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And here's my response:
> > ```
> > {
> >   "reServiceCode": "POEFSC",
> >   "eTag": "MjAyNC0xMi0xMlQxNDo1NDo0OS40MzQ4MDJa",
> >   "id": "ad1dc99c-3c51-4c0b-94eb-ace44d6be2f1",
> >   "modelType": "MTOServiceItemBasic",
> >   "moveTaskOrderID": "5a99af3f-e9ce-4973-a064-b1606e3d95e8",
> >   "mtoShipmentID": "113a4bc3-ad46-464c-a98f-396b65ea2872",
> >   "reServiceName": "International POE Fuel Surcharge",
> >   "status": "APPROVED"
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I just am not sure how I feel about us not returning the values that we have updated to confirm that they have been updated correctly. Here I updated the port, but I have no way of knowing it was successful (other than the `200`) in the response. Any thoughts on that? We could entertain having a `MTOServiceItemFuelSurcharge` response that contains that data, or add it to `MTOServiceItemBasic`.
> > But it's also not a hill I want to die on and if the gov is cool with the response lacking all the port info, then I don't mind approving.
> 
> Yeah I do see what you're saying. Do you know other service items return any kind of info in the response like when they do an update? The way this feature spelled it out, after sending this update they would subsequently query the MTO and see the port information in that payload.

There a few more types that we use for SIT, etc. You can check these others out:
```
definitions:
  MTOServiceItemBasic: # spectral oas2-unused-definition is OK here due to polymorphism
    $ref: 'definitions/prime/MTOServiceItemBasic.yaml'
  MTOServiceItemDestSIT: # spectral oas2-unused-definition is OK here due to polymorphism
    $ref: 'definitions/prime/MTOServiceItemDestSIT.yaml'
  MTOServiceItemDomesticCrating: # spectral oas2-unused-definition is OK here due to polymorphism
    $ref: 'definitions/prime/MTOServiceItemDomesticCrating.yaml'
  MTOServiceItemInternationalCrating: # spectral oas2-unused-definition is OK here due to polymorphism
    $ref: 'definitions/prime/MTOServiceItemInternationalCrating.yaml'
  MTOServiceItemOriginSIT: # spectral oas2-unused-definition is OK here due to polymorphism
    $ref: 'definitions/prime/MTOServiceItemOriginSIT.yaml'
  MTOServiceItemShuttle: # spectral oas2-unused-definition is OK here due to polymorphism
    $ref: 'definitions/prime/MTOServiceItemShuttle.yaml'
```

brianmanley-caci commented:
> MTOServiceItemShuttle

Thanks for the suggestion. I'm looking at what it would take to add a custom response type for international fuel surcharge now.

brianmanley-caci commented:
> > > Question - was able to update the service item, but here's my payload:
> > > ```
> > > {   
> > >    "id": "ad1dc99c-3c51-4c0b-94eb-ace44d6be2f1",   
> > >    "modelType": "UpdateMTOServiceItemInternationalPortFSC",   
> > >    "portCode": "PDX",   
> > >    "reServiceCode": "POEFSC" 
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > And here's my response:
> > > ```
> > > {
> > >   "reServiceCode": "POEFSC",
> > >   "eTag": "MjAyNC0xMi0xMlQxNDo1NDo0OS40MzQ4MDJa",
> > >   "id": "ad1dc99c-3c51-4c0b-94eb-ace44d6be2f1",
> > >   "modelType": "MTOServiceItemBasic",
> > >   "moveTaskOrderID": "5a99af3f-e9ce-4973-a064-b1606e3d95e8",
> > >   "mtoShipmentID": "113a4bc3-ad46-464c-a98f-396b65ea2872",
> > >   "reServiceName": "International POE Fuel Surcharge",
> > >   "status": "APPROVED"
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I just am not sure how I feel about us not returning the values that we have updated to confirm that they have been updated correctly. Here I updated the port, but I have no way of knowing it was successful (other than the `200`) in the response. Any thoughts on that? We could entertain having a `MTOServiceItemFuelSurcharge` response that contains that data, or add it to `MTOServiceItemBasic`.
> > > But it's also not a hill I want to die on and if the gov is cool with the response lacking all the port info, then I don't mind approving.
> > 
> > 
> > Yeah I do see what you're saying. Do you know other service items return any kind of info in the response like when they do an update? The way this feature spelled it out, after sending this update they would subsequently query the MTO and see the port information in that payload.
> 
> There a few more types that we use for SIT, etc. You can check these others out:
> 
> ```
> definitions:
>   MTOServiceItemBasic: # spectral oas2-unused-definition is OK here due to polymorphism
>     $ref: 'definitions/prime/MTOServiceItemBasic.yaml'
>   MTOServiceItemDestSIT: # spectral oas2-unused-definition is OK here due to polymorphism
>     $ref: 'definitions/prime/MTOServiceItemDestSIT.yaml'
>   MTOServiceItemDomesticCrating: # spectral oas2-unused-definition is OK here due to polymorphism
>     $ref: 'definitions/prime/MTOServiceItemDomesticCrating.yaml'
>   MTOServiceItemInternationalCrating: # spectral oas2-unused-definition is OK here due to polymorphism
>     $ref: 'definitions/prime/MTOServiceItemInternationalCrating.yaml'
>   MTOServiceItemOriginSIT: # spectral oas2-unused-definition is OK here due to polymorphism
>     $ref: 'definitions/prime/MTOServiceItemOriginSIT.yaml'
>   MTOServiceItemShuttle: # spectral oas2-unused-definition is OK here due to polymorphism
>     $ref: 'definitions/prime/MTOServiceItemShuttle.yaml'
> ```



> > MTOServiceItemShuttle
> 
> Thanks for the suggestion. I'm looking at what it would take to add a custom response type for international fuel surcharge now.

Ok, so I have added a new type MTOServiceItemInternationalFuelSurcharge that will include the service code and the port in the response. This should help validate that the port they sent was saved. I think this is a helpful solution.

brianmanley-caci commented:
> Ah, but while I'm thinking of it - please update the docs for the update service item endpoint on what is required when updating `PODFSC` & `POEFSC`. Just a brief run through of our process, what we require, etc, so we can inform HomeSafe when they start implementing this.

Idded some verbiage to the API documentation to help spell out how the port can be updated.

robot-mymove commented:
# Test coverage on server failed.
  **Last run: Fri 13 Dec 2024 01:57:59 PM UTC**
  Refer to [this report](https://output.circle-artifacts.com/output/job/1739ae81-71f0-42c0-b606-a00ad135cc1b/artifacts/0/test-results/gotest/app/go-coverage.html) to see coverage details for the server app.
  Documentation for test coverage can be found [at this link](https://transcom.github.io/mymove-docs/docs/about/development/test-coverage-reports).

robot-mymove commented:
# Test coverage on server failed.
  **Last run: Fri 13 Dec 2024 02:30:37 PM UTC**
  Refer to [this report](https://output.circle-artifacts.com/output/job/0c3943ca-3215-4839-9188-0c137feff1d8/artifacts/0/test-results/gotest/app/go-coverage.html) to see coverage details for the server app.
  Documentation for test coverage can be found [at this link](https://transcom.github.io/mymove-docs/docs/about/development/test-coverage-reports).

robot-mymove commented:
# Test coverage on server failed.
  **Last run: Fri 13 Dec 2024 06:13:48 PM UTC**
  Refer to [this report](https://output.circle-artifacts.com/output/job/bc2f2bf1-3fdf-456c-b9fc-92288c9ac51b/artifacts/0/test-results/gotest/app/go-coverage.html) to see coverage details for the server app.
  Documentation for test coverage can be found [at this link](https://transcom.github.io/mymove-docs/docs/about/development/test-coverage-reports).

robot-mymove commented:
# Test coverage on server failed.
  **Last run: Fri 13 Dec 2024 10:08:15 PM UTC**
  Refer to [this report](https://output.circle-artifacts.com/output/job/6de3acf3-f50c-4ebe-85ee-3454cee2d759/artifacts/0/test-results/gotest/app/go-coverage.html) to see coverage details for the server app.
  Documentation for test coverage can be found [at this link](https://transcom.github.io/mymove-docs/docs/about/development/test-coverage-reports).

robot-mymove commented:
# Test coverage on server failed.
  **Last run: Mon 16 Dec 2024 02:30:02 PM UTC**
  Refer to [this report](https://output.circle-artifacts.com/output/job/61f3e6cc-4270-409a-9e43-6c6691486f8f/artifacts/0/test-results/gotest/app/go-coverage.html) to see coverage details for the server app.
  Documentation for test coverage can be found [at this link](https://transcom.github.io/mymove-docs/docs/about/development/test-coverage-reports).

danieljordan-caci reviewed (approved):
Update response:
![Screenshot 2024-12-16 at 9 21 23 AM](https://github.com/user-attachments/assets/c0cac46b-1d19-4fa1-ba13-4c07b990dd90)

Shipment body in MTO fetch:
![Screenshot 2024-12-16 at 9 22 03 AM](https://github.com/user-attachments/assets/16e79d1e-c5a3-47d4-9d52-647335c07ba9)

LGTM! Thanks for all the changes.

danieljordan-caci commented on swagger-def/definitions/Port.yaml:
```diff
@@ -0,0 +1,95 @@
+    description: 'Port type A (Air), B (Border Crossing), S (Sea)'
+    enum:
+      - A
+      - P
```
oooo nice eagle eye!

brianmanley-caci replied:
Nice catch. It should be update now.
